test/e2e/clusterctl_upgrade.go

```diff
@@ -428,32 +431,55 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			Name:                   workloadClusterName,
 		}, input.E2EConfig.GetIntervals(specName, "wait-cluster")...)
 
-		expectedMachineCount := *controlPlaneMachineCount + calculateExpectedWorkerCount(ctx, managementClusterProxy.GetClient(), workloadClusterUnstructured, coreCAPIStorageVersion)
+		expectedWorkerCount := calculateExpectedWorkerCount(ctx, managementClusterProxy.GetClient(), workloadClusterUnstructured, coreCAPIStorageVersion)
+		expectedMachineCount := *controlPlaneMachineCount + expectedWorkerCount
 
-		By("Waiting for the machines to exist")
-		Eventually(func() (int64, error) {
-			var n int64
-			machineList := &unstructured.UnstructuredList{}
-			machineList.SetGroupVersionKind(schema.GroupVersionKind{
-				Group:   clusterv1.GroupVersion.Group,
-				Version: coreCAPIStorageVersion,
-				Kind:    "MachineList",
-			})
-			if err := managementClusterProxy.GetClient().List(
-				ctx,
-				machineList,
-				client.InNamespace(workloadClusterNamespace),
-				client.MatchingLabels{clusterv1.ClusterNameLabel: workloadClusterName},
-			); err == nil {
-				for _, m := range machineList.Items {
-					_, found, err := unstructured.NestedMap(m.Object, "status", "nodeRef")
-					if err == nil && found {
-						n++
+		if !input.SkipMachinePoolMachines {
```

Comment:
Shouldn't we at least always check for the `controlPlaneMachineCount then?

```suggestion
		expectedMachineCount := *controlPlaneMachineCount
		if !input.SkipMachinePoolMachines {
				expectedMachineCount += expectedWorkerCount
		}
```

and not have this if?